### ISSUE TITLE
Run prisma migrate command programmatically

### ISSUE THREAD

**Issue by dested** [prisma/prisma#4703]
## Problem

It would be nice to have a way to run the migrate programmatically. I am running into an issue where my api is using lambdas (serverless) and my database is only accessible from within an aws vpc. Because of this I cannot "execute commands" on deploy in an easy way. My current solution is to ssh into an ec2 that is within the vpc, execute the prisma commands, and then spin the ec2 back down. While this is working, it is far from ideal. If I was instead able to have a separate lambda that had all the migration data bundled with it I could execute the migration from a lambda (that is within the vpc). 

While this may seem like a specialized use case, databases that are not publicly accessible are common place, and solutions for migration that don't include running npm commands from the cli I think become critical.

## Solution
```
import {migrate} from '@prisma/cli';
migrate('up');
```

## Alternatives
Stated above, no great alternatives for common usecases.

**Comment by PhilippMolitor** [comment]
I had some thoughts on this feature and already shared them with @albertoperdomo from the Prisma team.
To me, a very essential feature that would be possible with this addition would be to run a custom seeder.
Imagine freshly deploying the app onto a server, and when it detects that there are 0 migrations in the database yet, it runs a custom function creating some essential entries.

Imagine something like this:

```typescript
const connection = new PrismaClient();

connection.$connect()
  .then(() => {
    const migrations = connection.$migrations();

    const isFreshInstall = migrations.databaseIsInitialized;

    if(migrations.pending > 0) {
      migrations.runPendingMigrations();
    } else if (migrations.schemaNotMatching) {
      throw new Error('server schema newer or not compatible with application schema');
    }

    if(isFreshInstall) {
      connections.mymodel.create({data: { ... }});
      console.log('database was seeded!');
    }
  });
```

**Comment by yardenshoham** [comment]
I'm packaging my app with `pkg` and I can't expect the environment to have node/npm/npx/yarn. How can I go about creating the sqlite file and initializing the database programmatically (basically `npx prisma migrate reset --preview-feature`) so in a new environment, the app could be run without any shell interaction?

**Comment by johnnyoshika** [comment]
Automatic migration in code would be really nice for me as well.

**Comment by AhmetHuseyinDOK** [comment]
I also need this feature for my projects too, love prisma, hope this will feature will be implemented soon

**Comment by thdxr** [comment]
Dug through the Rust code and I found a way to do this by directly talking to the `migration-engine` You can download this version here: https://binaries.prisma.sh/2.16.x/latest/rhel-openssl-1.0.x/migration-engine.gz and get it into your function either through a layer or directly.  Make sure it's chmod +x

The following typescript code will spawn it and message it with the right rpc command:

```
    const rpc = {
      jsonrpc: "2.0",
      method: "applyMigrations",
      params: ["./prisma/migrations"],
      id: 1,
    }
    const child = exec(
      `/opt/nodejs/migration-engine -d ./prisma/schema.prisma`,
      {
        env: {
          DATABASE_URL: Config.postgres.url(),
        },
      }
    )
    child.stdin?.setDefaultEncoding("utf-8")
    child.stdin?.write(JSON.stringify(rpc))
    child.stdin?.end()

    return new Promise<APIGatewayProxyResult>((resolve) => {
      child.stdout?.on("data", (data) => {
        child.kill("SIGKILL")
        resolve({
          statusCode: 200,
          body: data,
        })
      })
    })
```

**Comment by Jolg42** [comment]
@thdxr this is definitely not recommended for many reasons (undocumented, can break at any time, not intended to be used like this, manually picking binaries ...)

And I don't see how this is a easier/better than using the `prisma` CLI like this for example: 
```
import execa from 'execa'

await execa.command(`prisma migrate deploy --preview-feature`)
``` 

Why not calling the CLI like this in your case?

**Comment by thdxr** [comment]
It's not as simple as calling the cli. I need to pick the binary anyway to make sure it's bundled and then do a few additional steps as documented in this comment: https://github.com/prisma/prisma/issues/4703#issuecomment-669801976

I understand why this isn't recommended but after struggling to get the cli working properly in a lambda I found this to be easier

**Comment by Jolg42** [comment]
@thdxr Definitely interested in your feedback here 👍 

It actually made me think that we actually have a (currently undocumented) way to select which binaries needs to be installed for the CLI
https://github.com/prisma/docs/issues/1152

Happy to get your feedback if that helps or not 😃

**Comment by thdxr** [comment]
Thanks for that link I'm able to get it working by using PRISMA_CLI_BINARY_TARGETS with the following step.

1. Copy the following files into a layer or directly - the reason I'm explicit here is that Lambda has a 250mb limit and copying all the Prisma folders starts to run up against it
```
"node_modules/@prisma/client",
"node_modules/@prisma/engines/migration-engine-rhel-openssl-1.0.x",
"node_modules/@prisma/engines/query-engine-rhel-openssl-1.0.x",
"node_modules/@prisma/engines/package.json",
"node_modules/@prisma/engines/dist",
"node_modules/prisma",
"node_modules/.prisma",
```
There's more trimming that could be done but it requires this list get really long so would love to see a better file structure for it (eg I don't need to include the prisma studio frontend)

2. In your lambda function invoke it like this if you want to avoid a dependency on `execa` - `process.execPath` will find the node binary for you
```
execSync(
  `${process.execPath} /opt/nodejs/node_modules/prisma/build/index.js migrate deploy --preview-feature`,
  {
    env: {
      DATABASE_URL: "<connection-string>",
    },
  }
)
```

**Comment by tmokmss** [comment]
Hi, I also created a sample to show how to run prisma migrate on AWS Lambda.
https://github.com/aws-samples/prisma-lambda-cdk/blob/main/backend/migration-runner.ts

**Comment by PhilippMolitor** [comment]
Is this issue still under investigation? I saw this was a candidate for a short time @tomhoule .

For the current projects I create, this feature would simplify my deployments a lot.

**Comment by janpio** [comment]
Having this would also make problems like this issue less relevant/important: https://github.com/prisma/prisma/issues/7933

**Comment by tomhoule** [comment]
It is something we can consider, but this needs a lot of product definition work. What migrate works with internally is probably not what people would want from an SDK. Currently we perceive other work (cascading deletes for example) as more important, but exposing a migrate SDK is definitely not out of the question.

It would be super helpful If everyone in this issue could give as much detail as possible on their use cases and what kind of API they need. Feedback is very much appreciated :)

**Comment by throberto** [comment]
> It is something we can consider, but this needs a lot of product definition work. What migrate works with internally is probably not what people would want from an SDK. Currently we perceive other work (cascading deletes for example) as more important, but exposing a migrate SDK is definitely not out of the question.
> 
> It would be super helpful If everyone in this issue could give as much detail as possible on their use cases and what kind of API they need. Feedback is very much appreciated :)

That would be interesting to me. I use AWS Aurora + AWS Lambda and would like to generate databases for each registered customer. 

The idea is to add a trigger in a DynamoDB Streams table that when adding a new customer invokes this lambda function to generate a customer database. I hope I helped with this use case guys. 😄

**Comment by PhilippMolitor** [comment]
I am deploying either manually or with CI via SSH to VPS servers, or I deploy to Amazon EBS. In both cases it feels rather inconvenient to manually install the prisma cli to some instance and manually run the migrations.

I had a very pleasant experience while using knex'es migration tool, which has a programmatic API built in.
It has the following features that were relevant for me:

- Detect amount of pending changes as number (positive = outstanding migrations, negative = database schema version is newer than application version, 0 = no pending migrations). This allowed me to very flexibly handle any edge case that could occur while automatically migrating.
- Knex has a "hidden" lock table, which keeps track of which migrations are already applied. Also, this stops two instances from migrating the database at the same time in a race condition.
- Finally, a method to run any outstanding migration from the migrations folder.

For a very basic example of how the API could look I provided an example some entries above this one in this issue: https://github.com/prisma/prisma/issues/4703#issuecomment-751299151

I'd love to see some of the ideas that went into PrismaClient also going into the migration API, e.g. when running `prisma migrate dev` that all SQL migrations are automatically generated as code for later use, this way everything would be automatically embedded in a production build.

**Comment by revmischa** [comment]
I am trying to build a serverless application with AWS tools only, like CDK and the [sam-cdk-cli-beta](https://aws.amazon.com/blogs/compute/better-together-aws-sam-and-aws-cdk/). I wrote an [article](https://spiegelmock.com/2021/05/29/frameworkless-web-applications-aws-cdk/) that explains my approach and thinking for anyone interested.

Essentially I want to use CDK to deploy functions that access an Aurora database and be able to run migrations, seed, etc in my cloud environment.
I was able to achieve this by deploying the following construct:
```ts
export class MigrationScript extends Construct {
  constructor(scope: Construct, id: string, { vpc, db }: DatabaseScriptProps) {
    super(scope, id)

    // migrate script
    const func = new NodejsFunction(this, "migrate", {
      timeout: Duration.seconds(60), 
      memorySize: 1024,
      layers: [db.getPrismaLayerVersion()], // layer: https://github.com/jetbridge/lambda-layer-prisma-pg
      environment: {
        NODE_OPTIONS: "--enable-source-maps",
      },
      runtime: Runtime.NODEJS_14_X,
      bundling: {
        externalModules: db.getExternalModules(),
        // minify: true,
        commandHooks: {
          beforeInstall: () => [],
          afterBundling: () => [],
          beforeBundling: (_inputDir: string, outputDir: string): string[] => {
            return [
              // need to copy over migration files
              `cp -r "${__dirname}/../../../repo/prisma" "${outputDir}"`,
            ]
          },
        },
      },
      vpc,
    })

    // allow DB access
    db.connections.allowDefaultPortFrom(func)
    if (db.secret) {
      func.addEnvironment(DB_SECRET_ENV, db.secret.secretArn)
      db.secret.grantRead(func)
    }

    // TODO
    new CfnOutput(this, "Invoke migrations", {
      value: "aws lambda invoke .....",
    })
  }
}
```
And the hacked together migration function:
```ts
import { Migrate } from "@prisma/migrate"
import { ensureDatabaseExists } from "@prisma/migrate/dist/utils/ensureDatabaseExists"
import { printFilesFromMigrationIds } from "@prisma/migrate/dist/utils/printFiles"
import chalk from "chalk"
import { getDatabaseUrl } from "myapp-repo" // defined elsewhere

export const handler = async (): Promise<string> => {
  // load DB URL from secrets
  if (!process.env.DATABASE_URL) {
    const dbUrl = await getDatabaseUrl()
    process.env.DATABASE_URL = dbUrl
  }

  const schemaPath = "./prisma/schema.prisma"

  const migrate = new Migrate()
  const wasDbCreated = await ensureDatabaseExists("apply", true, schemaPath)
  if (wasDbCreated) {
    console.info() // empty line
    console.info(wasDbCreated)
  }

  const diagnoseResult = await migrate.diagnoseMigrationHistory({
    optInToShadowDatabase: false,
  })
  const listMigrationDirectoriesResult = await migrate.listMigrationDirectories()

  if (listMigrationDirectoriesResult.migrations.length > 0) {
    const migrations = listMigrationDirectoriesResult.migrations
    console.info(`${migrations.length} migration${migrations.length > 1 ? "s" : ""} found in prisma/migrations`)
  } else {
    console.info(`No migration found in prisma/migrations`)
  }

  const editedMigrationNames = diagnoseResult.editedMigrationNames
  if (editedMigrationNames.length > 0) {
    console.info(
      `${chalk.yellow("WARNING The following migrations have been modified since they were applied:")}
${editedMigrationNames.join("\n")}`
    )
  }

  const { appliedMigrationNames: migrationIds } = await migrate.applyMigrations()

  migrate.stop()

  console.info() // empty line
  if (migrationIds.length === 0) {
    return chalk.greenBright(`No pending migrations to apply.`)
  } else {
    return `The following migration${migrationIds.length > 1 ? "s" : ""} have been applied:\n\n${chalk(
      printFilesFromMigrationIds("migrations", migrationIds, {
        "migration.sql": "",
      })
    )}

${chalk.greenBright("All migrations have been successfully applied.")}`
  }
}
```

I use the generated secret from the database and provide the secret ARN to my functions. They set up the database connection like so:
```ts
import { GetSecretValueCommand, SecretsManagerClient } from "@aws-sdk/client-secrets-manager"
import { DB_SECRET_ENV } from "@jetkit/cdk"
import { PrismaClient } from "@prisma/client"

// running local dev?
const IS_SAM_LOCAL = process.env.SAM_LOCAL || process.env.AWS_ACCOUNT_ID === "123456789012"

const secretsClient = new SecretsManagerClient({})

// cached client
export let _prisma: PrismaClient

/**
 * Get database client.
 * Lazily creates client.
 */
export const getPrisma = async (): Promise<PrismaClient> => {
  if (_prisma) return _prisma

  // load connection URL from secrets if available
  if (process.env.DB_SECRET_ENV && !IS_SAM_LOCAL) {
    const url = await getDatabaseUrl()
    _prisma = new PrismaClient({ datasources: { db: { url } } })
  } else {
    // load from schema.prisma - DATABASE_URL env var
    _prisma = new PrismaClient()
  }

  return _prisma
}

/**
 * Load DB URL from secrets manager.
 */
export async function getDatabaseUrl(): Promise<string> {
  // if we're in lambda and trying to connect to the DB
  // we should be able to load the DB secret
  const dbSecretArn = process.env[DB_SECRET_ENV]
  if (!dbSecretArn) throw new Error(`Trying to access database in Lambda but ${DB_SECRET_ENV} is not defined`)

  // load secret and build DATABASE_URL
  let dbSecret
  try {
    dbSecret = await getSecret(dbSecretArn)
  } catch (e) {
    console.error(`Failed to load database secret ${DB_SECRET_ENV}`)
    throw e
  }
  if (!dbSecret) throw new Error(`Failed to load database secret ${DB_SECRET_ENV}`)
  return databaseSecretToUrl(dbSecret)
}

function databaseSecretToUrl(secret: Record<string, string>): string {
  const { dbname, engine, port, host, username, password } = secret
  return [`${engine}:/`, `${username}:${password}@${host}:${port}`, dbname].join("/")
}

// move me
async function getSecret(secretArn: string): Promise<Record<string, string> | undefined> {
  const getCmd = new GetSecretValueCommand({ SecretId: secretArn })
  const res = await secretsClient.send(getCmd)
  if (!res.SecretString) return undefined
  return JSON.parse(res.SecretString)
}
```

I plan to provide this functionality in a reusable form in https://github.com/jetbridge/jetkit-cdk/ and https://github.com/jetbridge/typescript-cdk-template if it is generally "correct" and useful. It would of course be nice if I could call `const status = migrate.deploy()`.
Also being able to use the aurora data API (https://github.com/prisma/prisma/issues/1964) would [make things a lot simpler](https://github.com/prisma/prisma/issues/1964#issuecomment-878284201).

My biggest [problem](https://github.com/prisma/prisma/discussions/8245) right now is trying to bundle lambda functions with the generated prisma client without the rest of the prisma deps which live in my layer.

I hope this is helpful. I hope to provide this functionality in a reusable form on npm once I get it all straightened out.

**Comment by tomhoule** [comment]
Hey @revmischa — this looks super interesting! The `@prisma/migrate` package is considered an internal implementation detail, the API can break without notice, let's get in touch (maybe through the public Prisma slack?) to see how your use case can be supported and how we can avoid breaking your project, because `@prisma/migrate` is probably not the solution. When we work on a programmatic API, it will be almost certainly be something else.

**Comment by revmischa** [comment]
> Hey @revmischa — this looks super interesting! The `@prisma/migrate` package is considered an internal implementation detail, the API can break without notice, let's get in touch (maybe through the public Prisma slack?) to see how your use case can be supported and how we can avoid breaking your project, because `@prisma/migrate` is probably not the solution. When we work on a programmatic API, it will be almost certainly be something else.

Sure I just posted in #general on slack. I started a PR to share my code in a reusable form: https://github.com/jetbridge/jetkit-cdk/pull/13

**Comment by supermacro** [comment]
While still in its research phase, our team is exploring creating an integration testing framework for prisma+mysql whereby:

- we set up a single mysql server
- for every integration test, set up a dedicated schema / db so that each test is isolated and has its own database state (mysql does not have a limit on the number of schemas / dbs on a single server)
- have a db seeding API to allow developers to scaffold the db with whatever data they need in order to run a particular test
- seeding will be deterministic and not random to ensure reliable tests
- run tests in parallel

---

Every integration test would have the following "setup" steps (eg before any action or expectations can be done):

- Create dedicated database / schema:

```sql
CREATE DATABASE <uniqueRandomlyGeneratedValue>;
```

- Run `npx prisma migrate deploy` on table `uniqueRandomlyGeneratedValue`

To run migrations, as of now, we would need to use `shelljs`

```typescript
const shell = require('shelljs');

const BASE_DB_URL = 'mysql://root:root@localhost:3306'

const migrateDb = (autogeneratedSchemaName: string): Promise<void> => {
  const command = [
    `DATABASE_URL=${BASE_DB_URL}/${autogeneratedSchemaName}`,
    'npx prisma migrate deploy',
  ].join(' ') 

  return new Promise((resolve, reject) => {
    shell.exec(command, (code, stdout, stderr) => {
      if (code !== 0) {
        reject(stderr || stdout || 'non-zero exit code')
      }

      resolve()
    });  
  })
}
```

- Seed db with whatever data the developer has provided, and then return that specific prisma instance

```typescript
// domain-events/seeds.ts

import { PrismaClient } from '@prisma/client'

export const forCaseCreated = async (): Promise<PrismaClient> => {
  const autoGeneratedDbName: string = generateDatabaseName()

  await migrateDb(autoGeneratedDbName)

  const client = new PrismaClient({
    datasources: {
      db: {
        url: generateConnectionUrl(autoGeneratedDbName)
      }
    }
  })

  await Promise.all([
    a,
    bunch,
   of,
   different,
   prisma,
   queries,
  ])  

  return client
}
```

That means that each test would have approximately 3 to 4 seconds (complete un-empirical guess .. just based off of my own experience) of "setup" time. But since tests would be running in parallel, I don't imagine this becoming an issue as more tests are added.

**Comment by revmischa** [comment]
Maybe for MySQL that's the best you can do but for a real database like PostgreSQL you can run each test in its own isolated transaction and roll back after each test. Tests can be run in parallel because the transactions are isolated and never committed. Any transactions committed inside the tests can use [sub-transactions](https://www.postgresql.org/docs/current/sql-savepoint.html) (`SAVEPOINT`) so that to them the fact it's wrapped in an "outer" transaction is transparent. 

A wonderful example of this in python is [pytest-flask-sqlalchemy](https://pypi.org/project/pytest-flask-sqlalchemy/). This would be the ideal and fastest approach, at least with Postgres.

[I sincerely hope nobody is using MySQL for new projects in 2021](https://spiegelmock.com/2014/10/19/mysql-vs-postgresql-and-why-you-care/).

**Comment by Jolg42** [comment]
Programmatic Migrate could also help when a user wants to change the connection url at runtime maybe like in https://github.com/prisma/prisma/issues/2443#issuecomment-899662851

**Comment by jamesdebnam** [comment]
> While still in its research phase, our team is exploring creating an integration testing framework for prisma+mysql whereby:
> 
>     * we set up a single mysql server
> 
>     * for every integration test, set up a dedicated schema / db so that each test is isolated and has its own database state (mysql does not have a limit on the number of schemas / dbs on a single server)
> 
>     * have a db seeding API to allow developers to scaffold the db with whatever data they need in order to run a particular test
> 
>     * seeding will be deterministic and not random to ensure reliable tests
> 
>     * run tests in parallel
> 
> 
> Every integration test would have the following "setup" steps (eg before any action or expectations can be done):
> 
>     * Create dedicated database / schema:
> 
> 
> ```sql
> CREATE DATABASE <uniqueRandomlyGeneratedValue>;
> ```
> 
>     * Run `npx prisma migrate deploy` on table `uniqueRandomlyGeneratedValue`
> 
> 
> To run migrations, as of now, we would need to use `shelljs`
> 
> ```ts
> const shell = require('shelljs');
> 
> const BASE_DB_URL = 'mysql://root:root@localhost:3306'
> 
> const migrateDb = (autogeneratedSchemaName: string): Promise<void> => {
>   const command = [
>     `DATABASE_URL=${BASE_DB_URL}/${autogeneratedSchemaName}`,
>     'npx prisma migrate deploy',
>   ].join(' ') 
> 
>   return new Promise((resolve, reject) => {
>     shell.exec(command, (code, stdout, stderr) => {
>       if (code !== 0) {
>         reject(stderr || stdout || 'non-zero exit code')
>       }
> 
>       resolve()
>     });  
>   })
> }
> ```
> 
>     * Seed db with whatever data the developer has provided, and then return that specific prisma instance
> 
> 
> ```ts
> // domain-events/seeds.ts
> 
> import { PrismaClient } from '@prisma/client'
> 
> export const forCaseCreated = async (): Promise<PrismaClient> => {
>   const autoGeneratedDbName: string = generateDatabaseName()
> 
>   await migrateDb(autoGeneratedDbName)
> 
>   const client = new PrismaClient({
>     datasources: {
>       db: {
>         url: generateConnectionUrl(autoGeneratedDbName)
>       }
>     }
>   })
> 
>   await Promise.all([
>     a,
>     bunch,
>     of,
>     different,
>     prisma,
>     queries,
>     written,
>     by,
>     the,
>     developer,
>   ])  
> 
>   return client
> }
> ```
> 
> That means that each test would have approximately 3 to 4 seconds (complete un-empirical guess .. just based off of my own experience) of "setup" time. But since tests would be running in parallel, I don't imagine this becoming an issue as more tests are added.

This looks really interesting, and like something I could use for my api - but how do you go about dropping in the new prisma client into your code when testing?

**Comment by m1212e** [comment]
I've got the problem that the engine folder is too big: `Resource handler returned message: "Unzipped size must be smaller than 262144000 bytes (Service: Lambda, Status Code: 400,`

How did you solve this problem?

**Comment by revmischa** [comment]
> I've got the problem that the engine folder is too big: `Resource handler returned message: "Unzipped size must be smaller than 262144000 bytes (Service: Lambda, Status Code: 400,`
> 
> How did you solve this problem?

This is what I'm currently using for 3.6.0
```ts
 const createBundleCommand = [
      // create asset bundle in docker
      "bash",
      "-c",
      [
        `mkdir -p ${layerDir}`,
        // install PRISMA_DEPS
        `cd ${layerDir} && HOME=/tmp npm install ${modulesToInstallArgs}`,
        // delete unneeded engines
        `rm -f ${engineDir}/introspection-engine* ${engineDir}/prisma-fmt*`,
        // sdk sux
        `rm -f ${sdkDir}/dist/libquery_engine*`,
        `rm -rf ${sdkDir}/dist/get-generators/*engine*`,
        // get rid of some junk
        `rm -rf ${nm}/prisma/build/public`,
        `rm -rf ${nm}/prisma/prisma-client/src/__tests__`,
        `rm -rf ${nm}/@types`,
      ].join(" && "),
    ]
```

**Comment by BALEHOK** [comment]
I'm setting up integration tests for my project (not hosted in AWS). Haven't found a nice and clean solution so far, decided to begin with what @supermacro suggested (building and executing a command with `shelljs`). Works pretty well for me.

**Comment by revmischa** [comment]
This is very handy if you're using CDK - https://docs.serverless-stack.com/constructs/Script
Runs a lambda function (source of mine is above) on stack update

**Comment by jcpage** [comment]
> This is what I'm currently using for 3.6.0
> 
> ```ts
>  const createBundleCommand = [
>       // create asset bundle in docker
>       "bash",
>       "-c",
>       [
>         `mkdir -p ${layerDir}`,
>         // install PRISMA_DEPS
>         `cd ${layerDir} && HOME=/tmp npm install ${modulesToInstallArgs}`,
>         // delete unneeded engines
>         `rm -f ${engineDir}/introspection-engine* ${engineDir}/prisma-fmt*`,
>         // sdk sux
>         `rm -f ${sdkDir}/dist/libquery_engine*`,
>         `rm -rf ${sdkDir}/dist/get-generators/*engine*`,
>         // get rid of some junk
>         `rm -rf ${nm}/prisma/build/public`,
>         `rm -rf ${nm}/prisma/prisma-client/src/__tests__`,
>         `rm -rf ${nm}/@types`,
>       ].join(" && "),
>     ]
> ```

You were able to delete the introspection and fmt binaries?  When I do this the migrate deploy command just tries to re-download them when it's run in the lambda (which fails because the lambda doesn't have open Internet access).

Is that because you aren't using the full migration CLI code?  I'm basically running this lambda: https://github.com/aws-samples/prisma-lambda-cdk/blob/main/backend/migration-runner.ts along with my own CDK code that surgically removes everything from @prisma & prisma packages - which unfortunately doesn't include a bunch of huge binaries due to above issue.

I did get the lambda ZIP file down t about 43MB which isn't great but still comfortably below the 50MB limit.  I noticed those enormous binaries aren't stripped, doing that would remove another ~10-15MB (but I can't due to the same issue, it checks the hash of the binaries and redownloads them!  Argh).  Of course removing them AND having a stripped migrate binary would put the lambda zip under 15MB total, which would be awesome.

**Comment by vnenkpet** [comment]
Is there actually any reason why Prisma doesn't have the same migration style as TypeORM, which just generates typescript/js migration classes with async `up` and `down` functions? That way migrations are run with access to the code and from within the code, which seems like the more logical way of doing things to me. 

Apart from the issue with serverless as others pointed out, I honestly can't imagine how I would use prisma for some more complex migrations that also require business logic to run over the data, which isn't a big deal when using TypeORM as you can just do whatever you want in the code while running the migration. Sometimes I can imagine you might also need to connect to an external API to manipulate the data while migrating and that's just something you will very hardly do with plain SQL.

I am now in process of deciding whether to use Prisma or TypeORM for the next project and this seems like a deal-breaker honestly, although in the rest of the features Prisma seems like a clear winner.

**Comment by Jolg42** [comment]
Hi @vnenkpet! I copied your comment to the related issue we have for this topic, [here](https://github.com/prisma/prisma/issues/4688#issuecomment-1041463071).

This issue is about adding programmatic access to the current CLI commands (deploy, reset ...).
Currently, the official API is the CLI which means a child process needs to be started (with execa for example) to execute the commands.

**Comment by ivanjeremic** [comment]
I use serverless nextjs (vercel) I would like to create models from an ui and migrate is there any chance to get this working?

**Comment by Jolg42** [comment]
@ivanjeremic Could you create a separate feature request with more details? It sounds different from what it is here.

**Comment by kajah** [comment]
Is there any update on this feature? I'm having to hack very painful workarounds running migrations with GCP Cloud Run.

**Comment by janpio** [comment]
Can you elaborate on these painful workarounds? That is the information we are thankful to get to be able to understand and prioritize these issues (of which we have _many_). Thanks.

**Comment by rhys-at-speer** [comment]
@AleksandrukTad 

You can use a jumpbox. Setup an EC2 instance in the same VPC which permits inbound SSH connections. Add a security group to both the jumpbox as well as the resource you're trying to access. Then connect to it via a reverse SSH tunnel:

```shell
ssh -N -L 5432:XXXX-auroradbcluster-XXXX.region.rds.amazonaws.com:5432 ubuntu@public-ipv4.amazonaws.com -i rds-cert.pem
``` 
At this point you can treat the service as if it was running on your local machine

**Comment by tmokmss** [comment]
Or if you like serverless migration infrastructure, here's the sample that uses Lambda for prisma migration. 
https://github.com/aws-samples/prisma-lambda-cdk

**Comment by kharann** [comment]
Any update on this feature? I'm having issues running prisma migrate with execa. I'm running ![testcontainers](https://github.com/testcontainers/testcontainers-node) and my vision is `yarn start` will automatically spin up a Postgres container and migrate. However, I'm currently having issues with running migrating after spinning up a testcontainer. Currently doing it in a hacky way with execa

**Comment by jonasgroendahl** [comment]
Anyone with a great solution?

Trying to do `prisma migrate deploy` in a lambda (created in Pulumi) 

I tried to approach to bundle a zip file with `prisma` and `@prisma` with only necessary files and then a lambda with same code here: https://github.com/aws-samples/prisma-lambda-cdk 

Currently getting this error:

```
INFO	prisma migrate deploy exited with error Command failed: /var/task/node_modules/prisma/build/index.js migrate deploy
Error: Migration engine exited.
```

**Comment by adri1wald** [comment]
This is the solution I've come up with to deploy migrations in an AWS  CDK deployment:

# services/api/migration/index.js
```js
// @ts-check
const response = require('cfn-response')
const prismaPath = require.resolve('prisma')

async function prisma(...arguments) {
  console.log(`Running \`prisma ${arguments.join(' ')}\``)
  const execaCommand = (await import('execa')).execaCommand
  const subprocess = execaCommand(`${prismaPath} ${arguments.join(' ')}`)
  subprocess.stdout?.pipe(process.stdout)
  subprocess.stderr?.pipe(process.stderr)
  await subprocess
}

async function handler(event, context) {
  console.log('Received event: ', JSON.stringify(event, null, 2))
  if (event.RequestType === 'Delete') return
  try {
    await prisma('migrate', 'deploy')
    response.send(event, context, response.SUCCESS, { message: 'ok' })
  } catch (error) {
    console.error(error)
    response.send(event, context, response.FAILED, { message: error.message })
  }
}

exports.handler = handler
```

# services/api/Dockerfile.migration
```dockerfile
FROM public.ecr.aws/lambda/nodejs:16

WORKDIR "${LAMBDA_TASK_ROOT}"

RUN npm install -g npm@latest

COPY package*.json tsconfig.json ./
# Need this before install I think
COPY prisma ./prisma

RUN npm config set ignore-engines true

# big hack to ensure same versions as our api without pulling in all dependencies
COPY log-lock-versions.js ./
RUN subset=$(node log-lock-versions.js prisma execa cfn-response) \
    && rm package*.json \
    && npm init -y \
    && npm i $subset

COPY migration ./

CMD [ "index.handler" ]
```

# services/api/log-lock-versions.js
```js
const packages = require('./package-lock.json').packages
const versions = process.argv
  .slice(2)
  .map((package) => {
    const key = `node_modules/${package}`
    const version = packages[key].version
    return `${package}@${version}`
  })
  .join(' ')
console.log(versions)
```

# lib/stacks/api.ts
```ts
import * as path from 'path'
import { Construct } from 'constructs'
import { Stack, StackProps, Duration, CustomResource } from 'aws-cdk-lib'
import * as cr from 'aws-cdk-lib/custom-resources'
import * as ec2 from 'aws-cdk-lib/aws-ec2'
import * as ecsPatterns from 'aws-cdk-lib/aws-ecs-patterns'
import * as rds from 'aws-cdk-lib/aws-rds'
import * as sm from 'aws-cdk-lib/aws-secretsmanager'
import * as lambda from 'aws-cdk-lib/aws-lambda'

export type ApiStackProps = StackProps & {
  /**
   * VPC into which to deploy the API
   */
  vpc: ec2.Vpc
  /**
   * Database backing the API
   */
  database: rds.DatabaseCluster
  /**
   * Name of the database
   * @default "postgres"
   */
  databaseName?: string
  /**
   * Database secrets
   */
  databaseSecrets: sm.ISecret
}

export class ApiStack extends Stack {
  public readonly api: ecsPatterns.ApplicationLoadBalancedFargateService

  constructor(scope: Construct, id: string, props: ApiStackProps) {
    super(scope, id, props)

    const dbName = props.databaseName ?? 'postgres'
    const dbHost = props.databaseSecrets.secretValueFromJson('host').toString()
    const dbPort = props.databaseSecrets.secretValueFromJson('port').toString()
    const dbUsername = props.databaseSecrets.secretValueFromJson('username').toString()
    const dbPassword = props.databaseSecrets.secretValueFromJson('password').toString()
    const dbUrl = `postgresql://${dbUsername}:${dbPassword}@${dbHost}:${dbPort}/${dbName}`

    /**
     * Database Migration CustomResource
     */

    const lambdaSG = new ec2.SecurityGroup(this, 'LambdaSG', {
      vpc: props.vpc,
      allowAllOutbound: true // default
    })

    const dbMigrationLambda = new lambda.DockerImageFunction(this, `MigrationLambda`, {
      vpc: props.vpc,
      securityGroups: [ lambdaSG ],
      vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_ISOLATED },
      code: lambda.DockerImageCode.fromImageAsset(path.join(__dirname, '..', '..', 'services', 'api'), {
        file: 'Dockerfile.migration'
      }),
      timeout: Duration.minutes(10),
      memorySize: 512,
      environment: {
        DATABASE_URL: dbUrl
      }
    })

    lambdaSG.connections.allowToDefaultPort(
      props.database, 'Allow connection between migration lambda and database'
    )

    const dbMigrationProvider = new cr.Provider(this, 'MigrationProvider', {
      onEventHandler: dbMigrationLambda
    })

    const dbMigrationResource = new CustomResource(this, 'MigrationResource', {
      serviceToken: dbMigrationProvider.serviceToken,
      properties: {
        refresh: getImageUri(dbMigrationLambda)
      }
    })

    // ... code to deploy API
  }
}

// TODO this is _super_ hacky, but...
// As long as this takes a DockerImageFunction, should be no problemo
function getImageUri(lambda: lambda.DockerImageFunction): string | undefined {
  const cfnFunction = lambda.node.defaultChild as lambda.CfnFunction
  const code = cfnFunction.code as lambda.CfnFunction.CodeProperty
  return code.imageUri
}
```

**Comment by adamjq** [comment]
I'm also running this with [testcontainers](https://github.com/testcontainers/testcontainers-node) like @kharann. My use case is running E2E Jest tests for a NestJS app.

The solution with [execa](https://github.com/sindresorhus/execa) didn't work for me ([I got this error](https://github.com/sindresorhus/execa/issues/465)) but I got it working using [async-spawn](https://github.com/ralphtheninja/await-spawn) which just wraps `spawn` in a promise:

```typescript
process.env.DATABASE_URL = '<DOCKER_DATABASE_URL>';

import spawn from 'await-spawn';
import { DockerComposeEnvironment, Wait } from 'testcontainers';
...

describe('E2E Tests', () => {
  beforeAll(async () => {
    
    // launch database with testcontainers using wait strategy to make sure it's up

    // run prisma migrations against docker
    const prismaOutput = await spawn('prisma', ['migrate', 'deploy']);
    console.log(`${prismaOutput}`);
  }
}
```

**Comment by LinusU** [comment]
@adamjq a little bit of topic, but did you set `"type": "module"` in your `package.json` file? That should fix the `execa` import.

The way we handled this on Lambda was to have a separate Lambda just to run the migrations. Since with the entire `prisma` cli package, the Lambda size grew to almost 50MB with just the `prisma` package, which is the limit.

This lambda is very simple though, it's just these three lines:

```js
import { execa } from 'execa'

export async function handler () {
  await execa('prisma', ['migrate', 'deploy'], { preferLocal: true, stdio: 'inherit' })
}
```

The `preferLocal` makes execa look in `node_modules/.bin` first, since we have `prisma` as a regular dependency. And the `stdio: 'inherit'` is so that the output makes its way to the CloudWatch log.

-----

I would have much rather have this in the same Lambda, as [the deploy-hook in Scandium](https://github.com/LinusU/scandium#deploy-hooks). Since that way we know that whenever a new version of our api is deployed to a specific environment, the migrations are run on that environments database before requests are being served by that version.

**Comment by Divaaan** [comment]
Hi all, I've been trying to follow the thread in the hopes of getting my script to run but I can't for some reason get the environment variable for my connection to the DB to work when spawning a child process using node as shown below:

```typescript
const bashSpawn: ChildProcessWithoutNullStreams = spawn('bash', [`${scriptsDirectory}/prisma-migrate.sh`], {
    env: {
        PATH: process.env.PATH,
        PRISMA_SCHEMA: resolve(__dirname, './../prisma/schema.prisma'),
        DATABASE_CONNECTION_URL: `sqlserver\://${process.env.DATABASE_HOST}\;database=db-${site}\;user=${process.env.DATABASE_USER}\;password=${process.env.DATABASE_PASSWORD}\;encrypt=true\;trustServerCertificate=true\;`
    }
})
```

I've tried escaping the `:` and `;` characters but nothing works. Even when echoing the environment variable it look correct but prisma doesn't like it for some reason. 

I've tried setting the process.env variable and passing it through but still nothing.

Command I'm trying to run from the script is: `npx prisma migrate dev --name init --schema "$PRISMA_SCHEMA"`

Any suggestions would be great and will post my resolution here as well once I manage to get it working.

**Comment by adamjq** [comment]
Thanks @LinusU, I tried `"type": "module"` but it didn't fix it. I'm trying to use [Ecmascript modules with Jest](https://jestjs.io/docs/ecmascript-modules).

Regarding your use case, I think the way you have it currently configured with separate lambdas is the way to go. Bundling the Prisma CLI with the function that serves your API requests will increase cold start times. If you do need access to the CLI binary you might want to consider using Lambda Layers.

The way I've configured migrations to run in [my projects](https://github.com/adamjq/production-ready-serverless-nestjs) is prior to function deployment, so something like:
```yaml
    steps:
      ...
      - run: yarn test # run unit tests
      - run: yarn test:e2e # run integration tests using testcontainers with database migrations
      - run: npx prisma migrate deploy # run database migrations from GitHub actions
      - run: npx serverless deploy # deploy lambda
```

If you need to run migrations from infrastructure in a VPC then your approach or the previously mentioned https://github.com/aws-samples/prisma-lambda-cdk look promising. `prisma-lambda-cdk` uses [execFile](https://github.com/aws-samples/prisma-lambda-cdk/blob/main/backend/migration-runner.ts#L2). If I was to use that I would replace `npx prisma migrate deploy` with `aws lambda invoke ...` in my CI/CD pipeline.

I agree with the above comments that being able to run migrations from code directly would be much simpler

**Comment by Divaaan** [comment]
Nevermind on my query above, I was an idiot 🤦🏼 
Port in my environment variable was incorrect!

**Comment by mjknight50** [comment]
This is a need for us as well.  
We would like to be able to programmatic generate the sql scripts from our prisma.schema instead of using the cli.

The reason being that in our docker builds, we don't want to give bash access on the container.  We don't want to allow exec commands.

**Comment by jlipps** [comment]
I would also like programmatic access to the Prisma CLI. My use case is that I am developing an app which will be packaged via `pkg` and shipped to users' machines. I want to use sqlite to store data internally for this app, and Prisma seemed like a convenient ORM for this purpose. I want to be able to do things like:

- check if a sqlite file exists in the user's data dir
- create one with the correct schemas if not
- run migrations when the user updates the app

**Comment by fmaclen** [comment]
In a similar vein to @jlipps, I'm working on an Electron desktop app that stores user data in SQLite which is created and seeded on the user's device, it also receives occasional migrations. 

Here's how I am currently handling [creation/migration](https://github.com/Canutin/desktop/blob/master/sveltekit/src/lib/helpers/prisma.ts#L42-L74) and [seeding](https://github.com/Canutin/desktop/blob/master/sveltekit/src/lib/helpers/prisma.ts#L109-L128) from within the app.

**Comment by CaseyHaralson** [comment]
I'd like to be able to call a function via javascript that creates the database if it doesn't exist, and/or applies missing migrations. This would allow me to keep my databases private and not have to run other infrastructure (VMs or whatever).

My problem is I'm trying to use CD with a database that can't be accessed from outside the network. This is enough of a problem that I switch back and forth on whether I should just switch to another ORM.

**Comment by sarink** [comment]
A simpler solution:
1. `npm i @prisma/migrate`
2. ```
const { MigrateDeploy } = require('@prisma/migrate/dist/commands/MigrateDeploy');

new MigrateDeploy().parse([]);
```
3. 🎊

**Comment by janpio** [comment]
Yeah, and because that is an internal, not offical API that might just break without any warning in a future version @sarink. This will work, but be aware that this is not guaranteed to be maintained.

**Comment by sarink** [comment]
@janpio I know :-/ I can only hope that before this inevitably breaks the prisma team comes to their sense and publishes an official way to do this. There is simply no good solution for serverless/lambda folks.

**Comment by CaseyHaralson** [comment]
For my serverless lambda setup, I ended up using the @tmokmss idea of packaging everything into a docker container and then running that inside a lambda function.

I have a github workflow that includes a step to run prisma generate, copies those files to the dockerfile location, and then serverless picks up that folder for deploy to AWS. The workflow then hits the lambda endpoint to get auto-migration, but you could hit the endpoint manually.

[Serverless deploy Github workflow](https://github.com/CaseyHaralson/pinkyring-server-template/blob/main/.github/workflows/serverless.deploy.yml)

[Serverless file](https://github.com/CaseyHaralson/pinkyring-server-template/blob/main/serverless.yml)

[Serverless DbMigration folder](https://github.com/CaseyHaralson/pinkyring-server-template/tree/main/serverless/dbmigration)

**Comment by sarink** [comment]
This will vary greatly depending on your project configuration, but I can outline at a high level how the above solution works.

This assumes a folder structure like:
```
+ node_modules
* package.json
* build-migrate-lambda.sh
- migrate
  * index.js
```

First, install dependencies we need
```
npm i @prisma/migrate @prisma/engines @vercel/ncc
```

In your repo, create a new file
```
// migrate/index.js

const { MigrateDeploy } = require('@prisma/migrate/dist/commands/MigrateDeploy');

export const handler = () => {
  new MigrateDeploy().parse([]);
};
```

In a build script, compile it for lambda, eg `build-migrate-lambda.sh` (run this from the root)
```
#!/usr/bin/env bash

# Run npm install with the PRISMA_CLI_BINARY_TARGETS env var set to the engine required for the lambda environment (this will ensure that the right migration engine gets installed into `node_modules/@prisma/engines`, which we'll copy later
PRISMA_CLI_BINARY_TARGETS=rhel-openssl-1.0.x npm install

# Compile with ncc, make sure to put it in an additional subfolder ("migrate") because the prisma migrate code is hardcoded to look for a ../package.json file
./node_modules/.bin/ncc build ./migrate/index.js -o ./migrate/dist/migrate/ -m

# Copy package.json to the dist dir
cp -v package*.json ./migrate/dist/

# Copy the migration engine to the dist dir
cp -v ./node_modules/@prisma/engines/migration-engine-rhel-openssl-1.0.x ./migrate/dist/

# Zip it to prepare for lambda deployment
zip -r migrate-lambda.zip ./migrate/dist/
```

Now you should be able to deploy a lambda from the `migrate/dist` directory with the handler `migrate/index.js`, and invoke it as a post-deploy step

**Comment by olalonde** [comment]
I run my tests in parallel and they each need an isolated database, so programmatic access to migrations would be quite important. In the meantime, my workaround is:

`prisma-migrate.ts`

```typescript
import { promisify } from "node:util";
import { exec as execCb } from "node:child_process";

const exec = promisify(execCb);

// TODO: re-write this when Prisma.io gets a programmatic migration API
// https://github.com/prisma/prisma/issues/4703
export async function prismaMigrate(databaseUrl: string): Promise<void> {
  // throws an error if migration fails
  const { stdout, stderr } = await exec("npx prisma migrate dev", {
    env: {
      ...process.env,
      DATABASE_URL: databaseUrl,
    },
  });
  // console.log(stdout)
  // console.log(stderr)
}
```

**Comment by septatrix** [comment]
I too currently miss this feature. One aspect I would like to see take into account is that it should be possible to override the database url from such an API.

**Comment by kanocarra** [comment]
+100 to this. I saw the light and finally moved off typeorm. Loving prisma but one good thing about typeorm was that you could get NestJS to run the migrations at server startup so you never had to worry about timings. We have RDS in a VPC that we ideally don't want to connect our CI/CD to so would love the ability to run post server startup.

**Comment by revmischa** [comment]
This is what I'm currently using to run migrations in AWS lambda: https://github.com/jetbridge/sst-prisma/blob/master/backend/src/db/migrationScript.ts

**Comment by nemanjahifolks** [comment]
> 

How to get rid of these warnings that cancel the migration on Vercel?

```bash
Update available 4.13.0 -> 4.15.0 
Run the following to update   
yarn add --dev prisma@latest  
yarn add @prisma/client@latest  
```

```ts
const { stdout, stderr } = await execAsync('npx prisma migrate deploy');
```

**Comment by janpio** [comment]
Please open a new discussion or issue @nemanjahifolks. Make sure to include the actual warnings and describe how they cancel a migration on Vercel. (Above I see an update notifier message, which is just informative and does not stop the execution of Prisma CLI.)

**Comment by bonustrack** [comment]
Any update on this issue?

**Comment by janpio** [comment]
If there is no update comment from the team, then there is no update.

**Comment by ixartz** [comment]
I just recently try Drizzle ORM for https://github.com/ixartz/Next-js-Boilerplate and they give the option to do the migration programmatically, it's a game changer in term of DX: you can run the migration automatically when the server start, you can run the migration before each database call, etc... You can almost forget there is a need to do the migration.

It would be great if Prisma team can prioritize this feature

**Comment by vetlevo** [comment]
Any update or roadmap of when this feature comes? 😊

**Comment by GeeWee** [comment]
I think Prisma has an example of how to force the esbuild to include other files [here](https://www.prisma.io/docs/orm/prisma-client/deployment/serverless/deploy-to-aws-lambda#loading-required-files)

**Comment by oskarwilliams** [comment]
I would really love if this was added as well. I have a database in a private VPC as well and the work around I have now to migrate our database is pretty horrible.

**Comment by oskarwilliams** [comment]
Our horrible solution is the command and entrypoint of our instance in ECS uses the cli to migrate our DB, and then follows that by starting our service. This means the running container also now includes things like `zx` in order to do this approach.

**Comment by kornicameister** [comment]
@oskarwilliams it is not so horrible, that's exactly approach I have been taking with python's equivalent of prisma on one of the previous projects.

Anyway, I was able to hack away my solution based on layer using technique I found in various of comments.
What I think I can add is that my use case is a bit unique because I am not really building the AWS CDK application but the library which forced me to found a way how to expose the library with everything that is required in a single installable bundle.

**Comment by pdesjardins90** [comment]
Here's what I came up with working with the AWS CDK: a migrator lambda that runs during deployment and fails the deployment if there's a migration error. It has to include the prisma CLI and the prisma client as well as your prisma schema and migrations. Here's the gist of it:

```typescript
// migrator-cdk.ts
import { CustomResource, IResource } from 'aws-cdk-lib'
import { Architecture, Runtime } from 'aws-cdk-lib/aws-lambda'
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs'
import { Provider } from 'aws-cdk-lib/custom-resources'
import { Construct } from 'constructs'

export function createMigrator(scope: Construct): IResource {
  const lambda = new NodejsFunction(scope, 'MigratorLambda', {
    description: 'Database migrator lambda',
    runtime: Runtime.NODEJS_22_X,
    architecture: Architecture.ARM_64,
    bundling: {
      forceDockerBundling: true,
      nodeModules: ['prisma', '@prisma/client'],
      environment: {
        PRISMA_CLI_BINARY_TARGETS: 'linux-arm64-openssl-3.0.x'
      },
      commandHooks: {
        beforeInstall: (inputDir: string, outputDir: string) => [
          `mkdir -p ${outputDir}/prisma`,
          `cp ${inputDir}/prisma/schema.prisma ${outputDir}/prisma`,
          `cp -r ${inputDir}/prisma/migrations ${outputDir}/prisma`
        ],
        beforeBundling: () => [],
        afterBundling: () => []
      }
    }
    // ... your other props
  })

  const { serviceToken } = new Provider(scope, 'MigratorCDKProvider', {
    onEventHandler: lambda
  })

  return new CustomResource(scope, 'MigratorCDKCustomResource', {
    resourceType: 'Custom::DatabaseMigrator',
    serviceToken,
    properties: {
      lastDeployedAt: new Date().toISOString() // forces migrator to run on all deployments
    }
  })
}
```

```typescript
// migrator-lambda.ts
import { CdkCustomResourceEvent, CdkCustomResourceResponse } from 'aws-lambda'
import { execFile } from 'child_process'
import { promisify } from 'util'

const execFileAsync = promisify(execFile)

const DATABASE_URL = '' // Retrieve from secrets somewhere

export async function handler(
  event: CdkCustomResourceEvent
): Promise<CdkCustomResourceResponse> {
  console.debug('Received request', event)
  const { RequestId, RequestType } = event
  switch (RequestType) {
    case 'Create': {
      console.info(`${RequestType} request: migrating database.`)
      await migrate()
      return { PhysicalResourceId: RequestId }
    }

    case 'Update': {
      console.info(`${RequestType} request: migrating database.`)
      await migrate()
      return { PhysicalResourceId: event.PhysicalResourceId }
    }

    case 'Delete': {
      console.info(`${RequestType} request: skipping migration.`)
      return { PhysicalResourceId: event.PhysicalResourceId }
    }

    default: {
      throw new Error(`Unexpected requestType: ${RequestType}`)
    }
  }
}

async function migrate(): Promise<void> {
  console.debug('Migrating database')
  const { stdout, stderr } = await execFileAsync(
    './node_modules/prisma/build/index.js',
    ['migrate', 'deploy'],
    {
      env: {
        ...process.env,
        DATABASE_URL
      }
    }
  )
  console.info({ text: stdout })
  if (stderr) {
    console.error({ text: stderr })
  }
}

```

Just have to make sure the lambda has access to your database (permissions, security groups, vpc, credentials, etc.) This will run on every CDK deployment and make sure your database is up-to-date before deploying the rest of your stack.

**Comment by LazerJesus** [comment]
so, to put a pin in this, this is notsomething offered by the prisma ecosystem. we're relying no hacks?!

**Comment by pdesjardins90** [comment]
> so, to put a pin in this, this is notsomething offered by the prisma ecosystem. we're relying no hacks?!

Yeah, until the feature gets added in the client itself

**Comment by virus2016** [comment]
Think this will be a game changer for some. Prioritise this one team 🤜🤛

**Comment by outragedhuman** [comment]
This is a big deal for us too. We want to be able to wipe our DB in between tests.

**Comment by revmischa** [comment]
> This is a big deal for us too. We want to be able to wipe our DB in between tests.

Maybe consider https://github.com/chax-at/transactional-prisma-testing/

**Comment by shansmith01** [comment]
I'm using SST and i would like a way to run migrations programatically also. At the moment I am force to run prisma migrate build in my start command eg

```"start": "npx prisma migrate deploy && react-router-serve ./build/server/index.js",```

I would like more control than that

**Comment by capaj** [comment]
this is the reason I stopped using prisma and went to drizzle

**Comment by delebash** [comment]
Looking at the tests folder I figured out how to run cli commands directly from code without exec.  See https://github.com/prisma/prisma/issues/13549#issuecomment-3181240170

**Comment by dispatchrabbi** [comment]
I'd really like this! I create a new database for each integration test, and I currently have to push the schema using `execSync('npx prisma db push')` which doesn't feel great. I'd love a programmatic way to call that from my test setup script instead.
